### PR TITLE
Modern error handling for `req_perform_iterative()` and `req_perform_parallel()`

### LINKED ISSUE #365: Update parallel and iterative error handling

### PR DESCRIPTION
**Pull request by hadley**
Fixes #365

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -50,7 +50,10 @@
   is now deprecated (#314).
 
 * `req_perform_parallel()` replaces `multi_req_perform()`. `multi_req_perform()`
-  is now deprecated (#314).
+  is now deprecated (#314). `req_perform_parallel()` has an new error handling
+  strategy matching `req_perform_iterative()` and `req_perform_sequential()`.
+  It will bubble up errors by default by you can choose an alterantive strategy
```
Comment (jonthegeek) [inline NEWS.md:55]:
```suggestion
  It will bubble up errors by default, but you can choose an alteranative strategy
```

File: R/iterate.R [hunk 2]
```diff
@@ -71,14 +71,24 @@
...
 #' @param progress Display a progress bar? Use `TRUE` to turn on a basic
 #'   progress bar, use a string to give it a name, or see [progress_bars] to
 #'   customise it in other ways.
 #' @param path Optionally, path to save the body of request. This should be
 #'   a glue string that uses `{i}` to distinguish different requests.
 #'   Useful for large responses because it avoids storing the response in
 #'   memory.
-#' @return A list of [response()]s.
+#' @return
+#' A list, at most length `max_reqs`, containing [response]s and possibly one
```
Comment (jonthegeek) [inline R/iterate.R:87]:
```suggestion
#' A list, at most length `max_reqs`, containing [response()]s and possibly one
```

File: R/iterate.R [hunk 3]
```diff
@@ -71,14 +71,24 @@
...
+#' error object, if `on_error` is `"return"` and one of the requests errors.
+#' If present, the error object will always be last element in the list.
```
Comment (jonthegeek) [inline R/iterate.R:89]:
```suggestion
#' If present, the error object will always be the last element in the list.
```

File: R/multi-req.R [hunk 4]
```diff
@@ -29,12 +23,21 @@
...
-#' @returns A list the same length as `reqs` where each element is either a
-#'   [response] or an `error`.
+#' @param on_error What should happen if one of the requests fails?
+#'
+#'   * `stop`, the default: stop iterating with an error.
+#'   * `return`: stop iterating, returning all the successful responses
+#'     received so far, as well as an error object for the failed request.
+#'   * `continue`: continue iterating, recording errors in the result.
+#' @return
+#' A list, the same length as `reqs`, containing [response]s and possibly
```
Comment (jonthegeek) [inline R/multi-req.R:33]:
```suggestion
#' A list, the same length as `reqs`, containing [response()]s and possibly
```

File: R/multi-req.R [hunk 5]
```diff
@@ -29,12 +23,21 @@
...
+#' error objects, if `on_error` is `"return"` or `"continue"` and one of the
+#' responses error. If `on_error` is `"return"` and it errors on the ith
```
Comment (jonthegeek) [inline R/multi-req.R:35]:
```suggestion
#' responses errors. If `on_error` is `"return"` and it errors on the ith
```

**Review by jonthegeek — APPROVED** [review]
Just a few typos. I like seeing [Prefer an enum](https://design.tidyverse.org/boolean-strategies.html) applied!

File: R/iterate.R (hunk 2 above)
Comment (hadley) [reply R/iterate.R:87]: I think that makes it harder to read the docs?

**Pull request merged**